### PR TITLE
Falsy should not replace instance

### PR DESCRIPTION
**Pull request by koresar**
A bug in my composables implementation was not caught. This test is missing.

### CODE REVIEW

**Comment by koresar** [comment]
I have just added one more commit to this PR. A missing case with `deepStaticProperties`.

Oh, wow. Tests fail. The reference implementation is missing `deepStaticProperties` feature. :)

**Comment by troutowicz** [comment]
> The reference implementation is missing deepStaticProperties feature. :)

It is?

https://github.com/stampit-org/stamp-specification/blob/master/examples/compose.js#L36
https://github.com/stampit-org/stamp-specification/blob/master/examples/compose.js#L59

File: test/priority-tests.js [hunk 1]
```diff
@@ -91,4 +91,19 @@ test('compose override priorities', nest => {
...
+    const deepInstance = { e: 'deep' };
+    const stamp = compose({
+      deepStaticProperties: {
+        d: deepInstance
+      }
+    });
+    const actual = stamp.d;
+    const notExpected = stamp.compose.deepStaticProperties.d;
+
+    assert.notEqual(actual, notExpected,
```
Comment (troutowicz) [inline test/priority-tests.js]: I don't see how this test would ever pass. The stamp has no `staticProperties` that would override the `deepStaticProperties`. `actual` and `notExpected` are the same here.

File: test/instance-replacement-tests.js [hunk 2]
```diff
@@ -24,6 +24,27 @@ test('Instance replacement', assert => {
   })();
 });
 
+test('Instance not replaced', assert => {
+  const newInstance = undefined;
+
+  compose({
+    initializers: [
+      () => {
+        return newInstance;
```
Comment (ericelliott) [inline test/instance-replacement-tests.js]: Please return `null` or `false`, instead.

File: test/priority-tests.js [hunk 3]
```diff
@@ -91,4 +91,19 @@ test('compose override priorities', nest => {
...
+      'deepStaticProperties should not be assigned from descriptor to stamp');
```
Comment (ericelliott) [inline test/priority-tests.js]: What? Why not?

Comment (koresar) [inline test/priority-tests.js]: Same reason we are deeply merging deepProperties into object instances.
```js
const nodejsNetSetup = compose({ deepStaticProperties: { net: { maxSockets: 10 } } });

const myStamp1 = compose(nodejsNetSetup);
myStamp1.net.maxSockets = 1000; // increasing max sockets

const myStamp2 = myStamp2.compose(...);
assert(myStamp2.net.maxSockets, 10); // FAILS
```

File: test/instance-replacement-tests.js (hunk 2 above)
Comment (koresar) [inline test/instance-replacement-tests.js]: I need to rename test to 
> 'Instance should not be replaced'

I return `undefined` to make sure it was not used as an object instance. This case was not covered.

**Comment by koresar** [comment]
> staticPropertyDescriptors are copied by assignment: `descriptor.propertyDescriptors = _.assign({}, descriptor1.propertyDescriptors, descriptor2.propertyDescriptors)`

@troutowicz @ericelliott I do not agree with this line in the specs. I missed that somehow earlier.
This fails:
```js
const nodejsNetSetup = compose({ deepStaticProperties: { net: { maxSockets: 10 } } });

const myStamp1 = compose(nodejsNetSetup);
myStamp1.net.maxSockets = 1000; // increasing max sockets for specific myStamp1 ONLY

const myStamp2 = myStamp2.compose(...);
assert(myStamp2.net.maxSockets, 10); // FAILS
```

The code above is highly unexpected to stamp users. I did not expect **deep** properties to be **shallow** assigned.

**Comment by troutowicz** [comment]
Hmm, your example works for me using the example implementation.

```js
import assert from 'assert';
import compose from './examples/compose';

const nodejsNetSetup = compose({ deepStaticProperties: { net: { maxSockets: 10 } } });

const myStamp1 = compose(nodejsNetSetup);
myStamp1.net.maxSockets = 1000; // increasing max sockets for specific myStamp1 ONLY

const myStamp2 = compose(nodejsNetSetup);
assert(myStamp2.net.maxSockets, 10); // SUCCEEDS
```

**Comment by koresar** [comment]
Your code is different from mine. Instead of
```js
const myStamp2 = compose(nodejsNetSetup);
```
There should have been 
```js
const myStamp2 = compose(myStamp1);
```

**Comment by troutowicz** [comment]
Ah, the problem is [here](https://github.com/stampit-org/stamp-specification/blob/master/examples/compose.js#L12-L13).

```js
const createStamp = ({
      methods, properties, deepProperties, propertyDescriptors, initializers,
      staticProperties, deepStaticProperties, staticPropertyDescriptors
    })=> {

  ...
```

This should be changed to:

```js
const createStamp = compose => {
  const {
    methods, properties, deepProperties, propertyDescriptors, initializers,
    staticProperties, deepStaticProperties, staticPropertyDescriptorsdesc
  } = merge({}, compose);

  ...
```

The merge would take care of any nested object properties.

**Comment by koresar** [comment]
Well, I think that the problem is here:
https://github.com/stampit-org/stamp-specification/blob/master/examples/compose.js#L36

Deep properties are shallow copied (`assign`) instead of deep cloned.

**Comment by troutowicz** [comment]
Yep, that's it. `deepProperties` is in the same boat.

https://github.com/stampit-org/stamp-specification/blob/master/examples/compose.js#L21

**Comment by koresar** [comment]
Oh, wow.

**Comment by ericelliott** [comment]
We need tests for both deep merge problems, and we should make the tests a lot more clear.

**Comment by koresar** [comment]
Do you have time on that? I can do today if I may.

File: test/instance-replacement-tests.js (hunk 2 above)
Comment (ericelliott) [inline test/instance-replacement-tests.js]: I completely overlooked `const newInstance = undefined;`. Please move it into the initializer function to make it more clear. I don't see any reason to put that in the outer closure scope.

**Comment by ericelliott** [comment]
> Do you have time on that? I can do today if I may.

I'm about to sleep. It's getting close to 1am here. =)

**Comment by koresar** [comment]
I merged a different in-progress branch to make CI test succeed.

I made the example implementation to follow the specs strictly (plus tests). I assume this sentence means that non-object are ignored.
> If the initializer returns a different **object**, it replaces the instance.

```js
if ( isObject(returnValue) ) { // <- added the isObject check here
  obj = returnValue;
}
```

Also, I implemented deep merging for deepProperties and deepStaticProperties. Plus tests of course.

File: examples/compose.js [hunk 4]
```diff
@@ -15,17 +16,19 @@ const createStamp = ({
 
   const assign = Object.assign;
 
-  const Stamp = function Stamp (options = {}) {
+  const Stamp = function Stamp (...args) {
```
Comment (ericelliott) [inline examples/compose.js]: Please change to `(options, ...args)`.

Comment (koresar) [inline examples/compose.js]: Will be implemented in #48

**Comment by koresar** [comment]
:dancer:

**Pull request merged**